dubbo-config/dubbo-config-api/src/main/java/org/apache/dubbo/config/AbstractConfig.java

```
@@ -557,32 +595,20 @@ public void refresh() {
-                        }
-                    } else if (MethodUtils.isStringMapSetter(method)) {
-                        String value = StringUtils.trim(compositeConfiguration.getString(extractPropertyName(getClass(), method)));
-                        if (StringUtils.isNotEmpty(value)) {
-                            Method getterMethod = MethodUtils.findGetterBySetter(getClass(), method);
-                            if (getterMethod != null) {
-                                @SuppressWarnings("unchecked")
-                                Map<String, String> map = (Map<String, String>) getterMethod.invoke(this);
-                                if (map == null) {
-                                    map = new HashMap<>();
-                                }
-                                map.putAll(URL.valueOf("?"+value).getParameters());
-                                method.invoke(this, map);
-                            }
-                        }
+                if (MethodUtils.isSetter(method)) {
+                    String value = StringUtils.trim(compositeConfiguration.getString(extractPropertyName(getClass(), method)));
+                    // isTypeMatch() is called to avoid duplicate and incorrect update, for example, we have two 'setGeneric' methods in ReferenceConfig.
+                    if (StringUtils.isNotEmpty(value) && ClassUtils.isTypeMatch(method.getParameterTypes()[0], value)) {
+                        method.invoke(this, ClassUtils.convertPrimitive(method.getParameterTypes()[0], value));
+                    }
+                } else if (isParametersSetter(method)) {
+                    String value = StringUtils.trim(compositeConfiguration.getString(extractPropertyName(getClass(), method)));
+                    if (StringUtils.isNotEmpty(value)) {
+                        Map<String, String> map = invokeGetParameters(getClass(), this);
```

I think it's better to use `MethodUtils.findGetterBySetter` instead, also we can't just ignore exceptions while invoking